blockly/src/utils/pathfinding/PathfindingLogic.java

```diff
@@ -0,0 +1,164 @@
+package utils.pathfinding;
+
+import com.badlogic.gdx.ai.pfa.DefaultGraphPath;
+import com.badlogic.gdx.ai.pfa.GraphPath;
+import core.Game;
+import core.level.Tile;
+import core.level.utils.Coordinate;
+import core.utils.Tuple;
+import java.util.*;
+
+/**
+ * PathfindingLogic is an abstract class that provides the basic structure for pathfinding
+ * algorithms.
+ *
+ * <p>This class maintains the open and closed sets, tracks the path, and provides methods to
+ * initialize the search, perform steps, and retrieve the final path.
+ *
+ * <p>It is designed to be extended by specific pathfinding algorithms, such as BFS or DFS.
+ *
+ * <p>It uses a deque for the open set, allowing for both FIFO and LIFO behavior based on the
+ * constructor parameter.
+ *
+ * @see PathfindingVisualizer
+ * @see Node
+ * @see systems.PathfindingSystem PathfindingSystem
+ */
+public abstract class PathfindingLogic {
+  /** The starting coordinate for the pathfinding search. */
+  protected final Coordinate startNode;
+
+  /** The ending coordinate for the pathfinding search. */
+  protected final Coordinate endNode;
+
+  private final Deque<Node> openSet = new ArrayDeque<>();
+  private final boolean fifo;
+  private final Set<Coordinate> closedSet = new HashSet<>();
+  private final List<Tuple<Node, TileState>> steps = new ArrayList<>();
+  private final Map<Coordinate, Coordinate> cameFrom = new HashMap<>();
+  private Coordinate lastParent = null;
+
+  /**
+   * Constructor for PathfindingLogic.
+   *
+   * @param startNode The starting coordinate for the pathfinding search.
+   * @param endNode The ending coordinate for the pathfinding search.
+   * @param fifo True if the open set should be treated as a FIFO queue, false for LIFO.
+   */
+  protected PathfindingLogic(Coordinate startNode, Coordinate endNode, boolean fifo) {
+    this.startNode = startNode;
+    this.endNode = endNode;
+    this.fifo = fifo;
+  }
+
+  /**
+   * Check if the coordinate is in the open set.
+   *
+   * <p>A node is in the open set if it has been added to the open set but not yet processed.
+   *
+   * @param coord The coordinate to check.
+   */
+  protected void addToOpenSet(Coordinate coord) {
+    if (lastParent != null && !cameFrom.containsKey(coord)) {
+      cameFrom.put(coord, lastParent);
+    }
+    Node node = new Node(coord);
+    steps.add(Tuple.of(node, TileState.CURRENT));
+    if (fifo) openSet.addLast(node);
+    else openSet.push(node);
```

Comment:
Ich sehe, wo Du hin willst. 

Aber den Code versteht in einem halben Jahr niemand mehr. Kannst Du nicht einfach passende Datenstrukturen einsetzen? Also eine kleine Fassade vor die Deque - einmal als BSF-Queue (mit `push`=hinten rein und `pop`=vorne raus) und einmal als DFS-Stack (mit `push`=oben drauf und `pop`=oben runter)? Dann kannst Du hier einmal im Konstruktor die richtige Struktur anlegen und danach immer `push` und `pop` nutzen, ohne nochmal ein Flag mitführen und abfragen zu müssen.